### PR TITLE
fix(entities) Remove process_condition and move to a processor

### PR DESCRIPTION
**Pull request by evanh**
The process_condition function can be done by a processor, merge it with
the existing time series processor. I merged it with the existing one so
that there wouldn't have to be any defensive programming around which one
would run first.

This also ports only the existing functionality, which is not handling all the cases. 
There are already open tasks to clean up datetimes in Snuba so I decided to consolidate
badness instead of entirely rewriting it for now.

restofthefire

### CODE REVIEW

File: snuba/query/processors/timeseries_processor.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+                                AnyExpression(),
+                            ),
+                        ),
+                    ]
+                ),
+                Param("literal", LiteralMatch(None, Any(str))),
+            ),
+        )
+
+    def time_expr(
```
Comment (fpacifici) [inline snuba/query/processors/timeseries_processor.py:83]: private please

File: snuba/datasets/entities/discover.py [hunk 2]
```diff
@@ -492,7 +492,7 @@ def get_query_processors(self) -> Sequence[QueryProcessor]:
             apdex_processor(columnset),
             failure_rate_processor(columnset),
             HandledFunctionsProcessor("exception_stacks.mechanism_handled", columnset),
-            TimeSeriesColumnProcessor({"time": "timestamp"}),
+            TimeSeriesProcessor(self.__time_group_columns, self.__time_parse_columns),
```
Comment (fpacifici) [inline snuba/datasets/entities/discover.py:495]: I think we should move the condition processing early on in the list of query processors. That processor essentially properly parses a datetime literal into a datetime object. All query processors should have access to the result.
I'd say you can just move the whole processor to the beginning of the list.

File: snuba/query/processors/timeseries_processor.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+                                        "column_name",
+                                        Or(
+                                            [
+                                                String(tc)
+                                                for tc in self.__time_parse_columns
+                                            ]
+                                        ),
+                                    ),
+                                ),
+                                AnyExpression(),
```
[the remark this answers is not in the document]
Comment (evanh) [reply snuba/query/processors/timeseries_processor.py]: I replaced this with the new TransformedColumn matcher.

File: snuba/clickhouse/query_dsl/accessors.py [hunk 4]
```diff
@@ -189,7 +190,7 @@ def get_time_range(
                 ),
             ),
             (
-                Column(None, String(timestamp_field)),
+                TransformedColumn(Column(None, String(timestamp_field))),
```
Comment (fpacifici) [inline snuba/clickhouse/query_dsl/accessors.py:193]: This is going to impact the query splitter. Could you please add a test to ensure that works properly? (snuba/tests/test_split.py). You should be able to add a case with a timestamp embedded in a function.

File: snuba/datasets/entities/discover.py [hunk 5]
```diff
@@ -446,7 +446,7 @@ def __init__(self) -> None:
         events_ro_storage = get_storage(StorageKey.EVENTS_RO)
         transactions_storage = get_storage(StorageKey.TRANSACTIONS)
 
-        self.__time_group_columns: Mapping[str, str] = {}
+        self.__time_group_columns: Mapping[str, str] = {"time": "timestamp"}
         self.__time_parse_columns = ("timestamp",)
```
Comment (fpacifici) [inline snuba/datasets/entities/discover.py:450]: You do not need these two fields anymore. Could you just pass the dictionary and the tuple directly to the initializer of the TimeSeriesProcessor ?

File: snuba/datasets/entities/errors.py [hunk 6]
```diff
@@ -76,25 +75,10 @@ def get_extensions(self) -> Mapping[str, QueryExtension]:
 
     def get_query_processors(self) -> Sequence[QueryProcessor]:
         return [
+            TimeSeriesProcessor(self.__time_group_columns, self.__time_parse_columns),
```
Comment (fpacifici) [inline snuba/datasets/entities/errors.py:78]: Same, you do not need the attributes, just pass the hardcoded values here.

File: snuba/datasets/entities/events.py [hunk 7]
```diff
@@ -94,25 +94,10 @@ def get_extensions(self) -> Mapping[str, QueryExtension]:
 
     def get_query_processors(self) -> Sequence[QueryProcessor]:
         return [
+            TimeSeriesProcessor(self.__time_group_columns, self.__time_parse_columns),
```
Comment (fpacifici) [inline snuba/datasets/entities/events.py:97]: Same, please hardcode the values here.

File: snuba/datasets/entities/outcomes.py [hunk 8]
```diff
@@ -55,20 +54,5 @@ def get_extensions(self) -> Mapping[str, QueryExtension]:
     def get_query_processors(self) -> Sequence[QueryProcessor]:
         return [
             BasicFunctionsProcessor(),
-            TimeSeriesColumnProcessor(self.__time_group_columns),
+            TimeSeriesProcessor(self.__time_group_columns, self.__time_parse_columns),
```
Comment (fpacifici) [inline snuba/datasets/entities/outcomes.py:57]: Same issue present in all the other entities.

File: snuba/query/matchers.py [hunk 9]
```diff
@@ -323,4 +323,36 @@ def match(self, node: AnyType) -> Optional[MatchResult]:
...
+    """
+    Matches a Column that has been wrapped in zero or more Functions. It recursively
+    checks the parameters of the FunctionCall to see if at least one of them is a Column
+    matching the provided column name.
+
+    Note that this makes no guarantees about the functions being called, it naively
+    goes through all the parameters and returns the first column match it finds.
+    """
+
+    column: Optional[Column] = None
```
Comment (fpacifici) [inline snuba/query/matchers.py:337]: I don't think this class is useful without providing a Column, so it should not be optional and it should not have a default.

Also I think we should allow a match that is more complex than a column:
itt would be great if we could match
`ifNull(arrayElement(tags.values, indexOf(tags.key, 'asdasd')), '')`
as `arrayElement(tags.values, indexOf(tags.key, 'asdasd'))`. 
Though I would be ok in keeping this last part for a separate PR.

File: snuba/query/matchers.py [hunk 10]
```diff
@@ -323,4 +323,36 @@ def match(self, node: AnyType) -> Optional[MatchResult]:
...
+
+    def match(self, node: AnyType) -> Optional[MatchResult]:
+        if not isinstance(node, (ColumnExpr, FunctionCallExpr)):
+            return None
+
+        if isinstance(node, ColumnExpr) and self.column is not None:
+            return self.column.match(node)
+
+        if isinstance(node, FunctionCallExpr):
```
Comment (fpacifici) [inline snuba/query/matchers.py:346]: I think we should have a list of allowed functions to be ignored.
`count(toStartOfDay(timestamp)) < 4` (Or any boolean condition).
Should never be confused with a timestamp condition, since, while the `timestamp` column is wrapped in a function that still returns a timestamp. The result is then wrapped into a function that does not have this property.

We could also provide default lists of allowed functions for common use like `TIMESTAMP_FUNCITONS`, `ARITHMETIC_FUNCTIONS`.

File: tests/test_api.py [hunk 11]
```diff
@@ -1474,6 +1476,7 @@ def test_split_query(self):
                                 self.base_time + timedelta(minutes=self.minutes)
                             ).isoformat(),
                             "orderby": "-timestamp",
+                            "granularity": 60,
```
Comment (fpacifici) [inline tests/test_api.py:1479]: Why were these added.

**Review by fpacifici — CHANGES_REQUESTED** [review]
Please see my comment on the matcher and around the query splitter.
Since there is some impact in making changes to get_time_range, I would consider moving the matcher and that change to their own PR, your call.

File: tests/test_api.py (hunk 11 above)
Comment (evanh) [reply tests/test_api.py:1479]: I just realized that I broke things with my change. I will fix and revert this.

File: snuba/query/matchers.py (hunk 10 above)
Comment (evanh) [reply snuba/query/matchers.py:346]: I will definitely add something to ignore boolean and conditional functions, since this shouldn't match those. But I'm less sure about providing a whitelist of functions. Clickhouse has more than 60 date functions alone, and that's not counting arithmetic functions, generic functions like ifNull etc. Your proposal also assumes that the match cares if the column type is preserved. While that is sometimes true, it's not always true.

I think it's better to not provide any typing guarantees in this function and document it (which I did). Trying to handle every edge case to ensure that the type is matched would be too much work for minimal gain.

File: snuba/clickhouse/query_dsl/accessors.py (hunk 4 above)
Comment (evanh) [reply snuba/clickhouse/query_dsl/accessors.py:193]: I removed this for now. It can be done separately.

File: snuba/query/matchers.py (hunk 9 above)
Comment (evanh) [reply snuba/query/matchers.py]: I'll leave it to another PR for this change. Might be as simple as changing `column: Column` to be `pattern: Pattern[Expression]`.

File: snuba/query/matchers.py (hunk 10 above)
Comment (fpacifici) [reply snuba/query/matchers.py:347]: I don't think I can agree with this approach. 
There are definitively a lot more combinations of functions that would change the type of the column we look for or change the semantics (see the example `count(toStartOfDay(timestamp))`  than functions that are supposed to return the same type (thus we can list). Since we do not have control around how the user composes the query, there would be more chances to match a function we should not match by mistake than ignore an expression we should not ignore.

So we would have false positives and false negatives, but it seems that without listing the allowed functions we would optimize to avoid false negatives making a larger mistake than if we optimized to avoid false positives.

Comment (evanh) [reply snuba/query/matchers.py:347]: OK well I think this new matcher needs to be thought out more, so I'm going to remove it and open it again in a separate PR. I don't want this to block the removal of process_condition.

File: snuba/query/processors/timeseries_processor.py [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+                        column_match,
+                        FunctionCallMatch(
+                            Or(
+                                [
+                                    String("toStartOfHour"),
+                                    String("toStartOfMinute"),
+                                    String("toDate"),
+                                ]
+                            ),
+                            (column_match, LiteralMatch(Any(str)),),
```
Comment (fpacifici) [inline snuba/query/processors/timeseries_processor.py:63]: `(column_match, LiteralMatch(Any(str)))`

**Review by fpacifici — APPROVED** [review]
